dotnet/src/Connectors/Connectors.Memory.SqlServer/SqlServerMapper.cs

```diff
@@ -1,102 +1,119 @@
+                    });
                 }
             }
         }
 
         return record;
 
-        static void SetValue(IDictionary<string, object?> storageModel, object record, PropertyModel property, object? value)
+        static void PopulateValue(SqlDataReader reader, PropertyModel property, object record)
         {
             try
             {
-                property.SetValueAsObject(record, value);
+                var ordinal = reader.GetOrdinal(property.StorageName);
+
+                if (reader.IsDBNull(ordinal))
+                {
+                    property.SetValueAsObject(record, null);
+                    return;
+                }
+
+                switch (property.Type)
+                {
+                    case var t when t == typeof(byte):
+                        property.SetValue(record, reader.GetByte(ordinal)); // TINYINT
+                        break;
+                    case var t when t == typeof(short):
+                        property.SetValue(record, reader.GetInt16(ordinal)); // SMALLINT
+                        break;
+                    case var t when t == typeof(int):
+                        property.SetValue(record, reader.GetInt32(ordinal)); // INT
+                        break;
+                    case var t when t == typeof(long):
+                        property.SetValue(record, reader.GetInt64(ordinal)); // BIGINT
+                        break;
+
+                    case var t when t == typeof(float):
+                        property.SetValue(record, reader.GetFloat(ordinal)); // REAL
+                        break;
+                    case var t when t == typeof(double):
+                        property.SetValue(record, reader.GetDouble(ordinal)); // FLOAT
+                        break;
+                    case var t when t == typeof(decimal):
+                        property.SetValue(record, reader.GetDecimal(ordinal)); // DECIMAL
+                        break;
+
+                    case var t when t == typeof(string):
+                        property.SetValue(record, reader.GetString(ordinal)); // NVARCHAR
+                        break;
+                    case var t when t == typeof(Guid):
+                        property.SetValue(record, reader.GetGuid(ordinal)); // UNIQUEIDENTIFIER
+                        break;
+                    case var t when t == typeof(byte[]):
+                        property.SetValueAsObject(record, reader.GetValue(ordinal)); // VARBINARY
+                        break;
+                    case var t when t == typeof(bool):
+                        property.SetValue(record, reader.GetBoolean(ordinal)); // BIT
+                        break;
+
+                    case var t when t == typeof(DateTime):
+                        property.SetValue(record, reader.GetDateTime(ordinal)); // DATETIME2
+                        break;
+
+#if NET
+                    case var t when t == typeof(TimeOnly):
```

Comment:
nice! This refactoring is going to make https://github.com/microsoft/semantic-kernel/issues/11165 trivial ;)

Reply:
That's part of what prompted this refactoring ;)